rs/boundary_node/rate_limits/canister/canister.rs

```diff
@@ -97,49 +120,53 @@ fn disclose_rules(args: DiscloseRulesArg) -> DiscloseRulesResponse {
     Ok(())
 }
 
-// TODO: adjust quota
 #[query(decoding_quota = 10000)]
 fn http_request(request: HttpRequest) -> HttpResponse {
     match request.path() {
-        "/metrics" => serve_metrics(ic_cdk::api::time() as i64, encode_metrics),
+        "/metrics" => with_metrics_registry(|registry| export_metrics_as_http_response(registry)),
         _ => HttpResponseBuilder::not_found().build(),
     }
 }
 
-fn periodically_poll_api_boundary_nodes(interval: u64) {
+fn periodically_poll_api_boundary_nodes(interval: u64, canister_api: Arc<dyn CanisterApi>) {
     let interval = Duration::from_secs(interval);
 
-    ic_cdk_timers::set_timer_interval(interval, || {
-        ic_cdk::spawn(async {
-            if let Ok(canister_id) = Principal::from_text(REGISTRY_CANISTER_ID) {
-                match call::<_, (Result<Vec<ApiBoundaryNodeIdRecord>, String>,)>(
-                    canister_id,
-                    REGISTRY_CANISTER_METHOD,
-                    (&GetApiBoundaryNodeIdsRequest {},),
-                )
-                .await
-                {
-                    Ok((Ok(api_bn_records),)) => {
-                        API_BOUNDARY_NODE_PRINCIPALS.with(|cell| {
-                            *cell.borrow_mut() =
-                                HashSet::from_iter(api_bn_records.into_iter().filter_map(|n| n.id))
-                        });
-                    }
-                    Ok((Err(err),)) => {
-                        ic_cdk::println!("Error fetching API boundary nodes: {}", err);
-                    }
-                    Err(err) => {
-                        ic_cdk::println!("Error calling registry canister: {:?}", err);
-                    }
+    ic_cdk_timers::set_timer_interval(interval, move || {
+        let canister_api = canister_api.clone();
+        ic_cdk::spawn(async move {
+            let canister_id = Principal::from(REGISTRY_CANISTER_ID);
+            match call::<_, (Result<Vec<ApiBoundaryNodeIdRecord>, String>,)>(
+                canister_id,
+                REGISTRY_CANISTER_METHOD,
+                (&GetApiBoundaryNodeIdsRequest {},),
+            )
+            .await
+            {
+                Ok((Ok(api_bn_records),)) => {
+                    // Set authorized readers of the rate-limit config.
+                    canister_api.set_api_boundary_nodes_principals(
+                        api_bn_records.into_iter().filter_map(|n| n.id).collect(),
+                    );
+                    // Update metrics.
+                    let current_time = ic_cdk::api::time() as i64;
+                    LAST_SUCCESSFUL_REGISTRY_POLL_TIME.with(|cell| {
+                        cell.borrow_mut().set(current_time);
+                    });
+                }
+                Ok((Err(err),)) => {
+                    ic_cdk::println!("Error fetching API boundary nodes: {}", err);
+                }
+                Err(err) => {
+                    ic_cdk::println!("Error calling registry canister: {:?}", err);
```

Comment:
We should also have a metric that just counts the different type of responses (e.g., success, failed-fetch, failed-call).

Reply:
I don't know if we can track `query` calls in metrics as those can't change the state.

Reply:
or You meant just for this registry fetch call?

Reply:
I mean this very specific timer. That instead of logging, we should just have a metric similar to the `CANISTER_API_CALLS_COUNTER`, but maybe `TIMER_CALLS_COUNTER` that just tracks what the result of each timer execution was.

Reply:
done